containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java

```diff
@@ -0,0 +1,1233 @@
+		File imageFile = m_targetfolder.append(IMAGE_ID_FILE).toFile();
+
+
+		//This must all happen under Lock
+		synchronized (m_lockObjectManagedInit) {
+			// pathToImageIdMap is only changed under lockObjectManagedInit, so accessing
+			// without lockObject is fine
+			String imageId = m_targetpathToImageIdMap.getOrDefault(m_targetfolder, "");
+
+			// No need to lock m_lockObject, as long m_copyingMap is not accessed
+			m_copyingMap = m_targetpathToCopyingMap.get(m_targetfolder);
+			if (m_copyingMap == null) {
+				m_copyingMap = new HashMap<>();
+				m_targetpathToCopyingMap.put(m_targetfolder, m_copyingMap);
+			}
+
+			m_copyiedList = getCopiedSet(m_targetfolder);
+
+			//Read from file if possible
+			if(imageId.isEmpty() && imageFile.exists()) {
+				try (BufferedReader bufferReader = new BufferedReader(new FileReader(imageFile));) {
+					imageId = bufferReader.readLine();
+				}
+			}
+
+			if (!dockerImageId.equals(imageId)) {
+				// if image id has changed...all bets are off
+				// and we must reload all directories
+
+				// Check if someone else is copying - otherwise wait.
+				while(true) {
+					Job j = null;
+					synchronized (m_lockObject) {
+						if (m_copyingMap.isEmpty()) {
+							m_copyiedList.clear();
+							break;
+						}
+						j = m_copyingMap.values().iterator().next();
+					}
+					m_monitor.subTask("Wainting for XY to finish");
+					j.join(0, m_monitor);
+
+				}
+
+				// TODO Clean up directory;
+
+				imageFile.getParentFile().mkdirs();
+				try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(imageFile));) {
+					bufferedWriter.write(dockerImageId);
+					bufferedWriter.newLine();
+				}
+			}
+			m_targetpathToImageIdMap.put(m_targetfolder, dockerImageId);
+		}
+	}
+
+	/**
+	 * Write the state of the mirror to the disk to be loaded when eclipse is
+	 * started the next time.
+	 */
+	private void saveMirrorState() {
+		if (!m_mirror) {
+			throw new RuntimeException("This should only be called in mirror-mode");
+		}
+
+		if (m_copyiedList != null) {
+			synchronized (m_lockObject) {
+				File dirFile = m_targetfolder.append(COPY_STATE_FILE).toFile();
+				Set<String> writelist = m_copyiedList.stream().map(x -> x.toString()).collect(Collectors.toSet());
+				try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(dirFile))) {
+					oos.writeObject(writelist);
+				} catch (IOException e) {
+					Activator.logErrorMessage("Failed to save state", e);
+				}
+			}
+		}
+
+	}
+
+	/**
+	 * Check if a path was already copied to the host
+	 *
+	 * @param path The path to check
+	 * @return true if it was already copied
+	 */
+	private boolean checkIfAlreadyCopyied(Path path)
+	{
+
+		if (!m_mirror) {
+			throw new RuntimeException("This must only called in mirror-mode");
+		}
+
+		synchronized (m_lockObjectManagedInit) {
+			String dockerImageId = getImageId();
+			if (!dockerImageId.equals(m_targetpathToImageIdMap.get(m_targetfolder))) {

```

Comment:
Nope. Changed the Exception to "Mirroring multiple Images to the same folder is not possible". This is a bug at the call-site.

Reply:
Fine, but again, you need to translate all your exception messages except for cases of "this should never happen".  In most cases of should never happen unless the caller screws up, there should be an assert to guard against it rather than throwing an exception or else make it impossible for the caller to screw up.

Reply:
It's an assertions now. :) - As I'm not a Java-developer, I wasn't really aware of then -> Thanks. :)